src/goto-instrument/contracts/contract_wrangler.cpp

```diff
@@ -0,0 +1,391 @@
+  // Mangle loop contract function by function.
+  // We match function with a regex. There should one and only one function
+  // with name matched by the regex.
+  for(const auto &function_entry : this->functions)
+  {
+    bool function_found = 0;
+
+    for(const auto &function : goto_model.goto_functions.function_map)
+    {
+      // We find a matched function.
+      if(std::regex_match(function.first.c_str(), function_entry.first))
+      {
+        if(function_found)
+          throw deserialization_exceptiont(
+            "function regex" + function_entry.second.regex_str +
+            " match more than one function");
+
+        function_found = 1;
+
+        // Mangle all loop contracts in the function.
+        for(const auto &loop_entry : function_entry.second.loop_contract)
+        {
+          mangle(loop_entry, function.first);
+        }
+      }
+    }
+
+    if(!function_found)
+      throw deserialization_exceptiont(
+        "function " + function_entry.second.regex_str + " match no function");
+  }
+}
+
+void contract_wranglert::add_builtin_pointer_function_symbol(
+  std::string function_name,
+  const unsigned num_of_args)
+{
+  // Already exist.
+  if(symbol_table.lookup(CPROVER_PREFIX + function_name) != nullptr)
+    return;
+
+  code_typet::parameterst parameters;
+  for(unsigned i = 0; i < num_of_args; i++)
+  {
+    parameters.emplace_back(pointer_typet(signedbv_typet(32), 32));
+  }
+  symbolt fun_symbol(
+    CPROVER_PREFIX + function_name, code_typet(parameters, empty_typet()), "c");
+  symbol_table.add(fun_symbol);
+}
+
+void contract_wranglert::substitute_symbol(
+  exprt &expr,
+  const std::map<std::string, exprt> &symbol_map)
+{
+  for(depth_iteratort it = expr.depth_begin(); it != expr.depth_end(); it++)
+  {
+    if(it->id() != ID_symbol)
+      continue;
+    for(auto &symbol_entry : symbol_map)
+    {
+      if(it->get(ID_identifier) == symbol_entry.first)
+      {
+        exprt symbol(symbol_entry.second);
+        it.mutate().swap(symbol);
+        break;
+      }
+    }
+  }
+}
+
+void contract_wranglert::mangle(
+  const loop_contract_clauset &loop_contract,
+  const irep_idt &function_id)
+{
+  // Loop contracts mangling consists of four steps.
+  //
+  // 1. Construct a fake function with a fake loop that contains the loop
+  //    contracts for parsing the contracts.
+  // 2. Parse the fake function and extract the contracts as exprt.
+  // 3. Substitute symbols in the extracted exprt with the symbols in the
+  //    symbol table of the goto model according to the symbol map provided by
+  //    the user.
+  // 4. Typecheck all contract exprt.
+  // 5. Annotate all contract exprt to the corresponding loop.
+
+  // Construct a fake function to parse.
+  // void function_id()
+  // {
+  //    while(1==1)
+  //    __CPROVER_assigns(assigns_str) // optional
+  //    __CPROVER_loop_invariant(invariants_str)
+  //    __CPROVER_decreases(decreases_str) // optional
+  // }
+  ::std::ostringstream pr;
```

Comment:
I might know where this is coming from, but ... please use `std::` instead of `::std::`.